// @antv/f2/lib/components/guide/views/Rect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tslib = require("tslib");
var _fEngine = require("@antv/f-engine");
var _util = require("@antv/util");
var _default = exports.default = function _default(props, context) {
  var _a = props.theme,
    theme = _a === void 0 ? {} : _a;
  var _b = (0, _util.deepMix)((0, _tslib.__assign)({}, theme.rect), props),
    points = _b.points,
    style = _b.style,
    animation = _b.animation,
    offsetX = _b.offsetX,
    offsetY = _b.offsetY;
  var checkNaN = points.some(function (d) {
    return isNaN(d.x) || isNaN(d.y);
  });
  if (checkNaN) return null;
  var start = points[0] || {};
  var end = points[1] || {};
  var offsetXNum = context.px2hd(offsetX);
  var offsetYNum = context.px2hd(offsetY);
  var posX = Math.min(start.x, end.x) + (offsetXNum || 0);
  var posY = Math.min(start.y, end.y) + (offsetYNum || 0);
  return (0, _fEngine.jsx)("group", null, (0, _fEngine.jsx)("rect", {
    style: (0, _tslib.__assign)({
      x: posX,
      y: posY,
      width: Math.abs(end.x - start.x),
      height: Math.abs(start.y - end.y)
    }, style),
    animation: animation
  }));
};
