// @antv/f2/lib/components/point/withPoint.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tslib = require("tslib");
var _fEngine = require("@antv/f-engine");
var _geometry = _interopRequireDefault(require("../geometry"));
var _default = exports.default = function _default(View) {
  return /** @class */function (_super) {
    (0, _tslib.__extends)(Point, _super);
    function Point() {
      return _super !== null && _super.apply(this, arguments) || this;
    }
    Point.prototype.getDefaultCfg = function () {
      return {
        geomType: 'point'
      };
    };
    Point.prototype.render = function () {
      var props = this.props;
      var coord = props.coord;
      var records = this.mapping();
      var clip = this.getClip();
      return (0, _fEngine.jsx)(View, (0, _tslib.__assign)({}, props, {
        coord: coord,
        records: records,
        clip: clip
      }));
    };
    return Point;
  }(_geometry.default);
};
