// @antv/f2/es/components/guide/index.js
import withGuide from './withGuide';
import TextGuideView from './views/Text';
import PointGuideView from './views/Point';
import LineGuideView from './views/Line';
import ArcGuideView from './views/Arc';
import RectGuideView from './views/Rect';
import ImageGuideView from './views/Image';
import TagGuideView from './views/Tag';
import LottieGuideView from './views/Lottie';
import PolylineGuideView from './views/Polyline';
var DefaultGuideView = function DefaultGuideView() {
  return null;
};
var TextGuide = withGuide(TextGuideView);
var PointGuide = withGuide(PointGuideView);
var LineGuide = withGuide(LineGuideView);
var ArcGuide = withGuide(ArcGuideView);
var RectGuide = withGuide(RectGuideView);
var ImageGuide = withGuide(ImageGuideView);
var TagGuide = withGuide(TagGuideView);
var LottieGuide = withGuide(LottieGuideView);
var PolylineGuide = withGuide(PolylineGuideView);
export default withGuide(DefaultGuideView);
export { withGuide, TextGuide, PointGuide, ArcGuide, LineGuide, RectGuide, ImageGuide, TagGuide, LottieGuide, PolylineGuide };
