// @antv/f2/lib/components/scrollBar/vertical.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tslib = require("tslib");
var _fEngine = require("@antv/f-engine");
var _default = exports.default = function _default(props, context) {
  var coord = props.coord,
    range = props.range,
    position = props.position,
    layout = props.layout,
    style = props.style,
    background = props.background,
    barStyle = props.barStyle;
  var top = coord.top,
    height = coord.height;
  var left = layout.left,
    width = layout.width;
  var _a = (range === null || range === void 0 ? void 0 : range.y) || (range === null || range === void 0 ? void 0 : range.x),
    start = _a[0],
    end = _a[1];
  var barTop = height * start;
  var barHeight = height * (end - start);
  return (0, _fEngine.jsx)("group", {
    style: (0, _tslib.__assign)({
      display: 'flex',
      top: top,
      left: position === 'left' ? left - context.px2hd('8px') : left + width
    }, style)
  }, (0, _fEngine.jsx)("line", {
    style: (0, _tslib.__assign)({
      position: 'absolute',
      top: 0,
      left: 0,
      width: 0,
      height: height,
      stroke: 'rgba(202, 215, 239, .2)',
      lineCap: 'round',
      lineWidth: '8px'
    }, background)
  }), (0, _fEngine.jsx)("line", {
    style: (0, _tslib.__assign)({
      position: 'absolute',
      top: barTop,
      width: 0,
      height: barHeight,
      stroke: 'rgba(202, 215, 239, .5)',
      lineCap: 'round',
      lineWidth: '8px'
    }, barStyle)
  }));
};
