// @antv/f2/lib/components/sunburst/sunburstView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _fEngine = require("@antv/f-engine");
var _default = exports.default = function _default(props) {
  var coord = props.coord,
    node = props.node,
    onClick = props.onClick;
  var children = node.children;
  var _a = coord.center,
    x = _a.x,
    y = _a.y;
  var _renderNodes = function renderNodes(nodes) {
    return (0, _fEngine.jsx)("group", null, nodes.map(function (node) {
      var xMin = node.xMin,
        xMax = node.xMax,
        yMin = node.yMin,
        yMax = node.yMax,
        color = node.color,
        children = node.children;
      return (0, _fEngine.jsx)("group", {
        onClick: onClick
      }, (0, _fEngine.jsx)("sector", {
        attrs: {
          cx: x,
          cy: y,
          lineWidth: '1px',
          stroke: '#fff',
          startAngle: "".concat(xMin, " rad"),
          endAngle: "".concat(xMax, " rad"),
          r0: yMin,
          r: yMax,
          fill: color
        }
      }), children && children.length ? _renderNodes(children) : null);
    }));
  };
  return _renderNodes(children);
};
