// @antv/f2/lib/deps/f2-scale/src/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Category", {
  enumerable: true,
  get: function get() {
    return _base2.default;
  }
});
Object.defineProperty(exports, "Identity", {
  enumerable: true,
  get: function get() {
    return _index.default;
  }
});
Object.defineProperty(exports, "Linear", {
  enumerable: true,
  get: function get() {
    return _linear.default;
  }
});
Object.defineProperty(exports, "Log", {
  enumerable: true,
  get: function get() {
    return _log.default;
  }
});
Object.defineProperty(exports, "Pow", {
  enumerable: true,
  get: function get() {
    return _pow.default;
  }
});
Object.defineProperty(exports, "Quantile", {
  enumerable: true,
  get: function get() {
    return _quantile.default;
  }
});
Object.defineProperty(exports, "Quantize", {
  enumerable: true,
  get: function get() {
    return _quantize.default;
  }
});
Object.defineProperty(exports, "Scale", {
  enumerable: true,
  get: function get() {
    return _base.default;
  }
});
Object.defineProperty(exports, "Time", {
  enumerable: true,
  get: function get() {
    return _time2.default;
  }
});
Object.defineProperty(exports, "TimeCat", {
  enumerable: true,
  get: function get() {
    return _time.default;
  }
});
Object.defineProperty(exports, "getScale", {
  enumerable: true,
  get: function get() {
    return _factory.getScale;
  }
});
Object.defineProperty(exports, "getTickMethod", {
  enumerable: true,
  get: function get() {
    return _index2.getTickMethod;
  }
});
Object.defineProperty(exports, "registerScale", {
  enumerable: true,
  get: function get() {
    return _factory.registerScale;
  }
});
Object.defineProperty(exports, "registerTickMethod", {
  enumerable: true,
  get: function get() {
    return _index2.registerTickMethod;
  }
});
var _base = _interopRequireDefault(require("./base"));
var _base2 = _interopRequireDefault(require("./category/base"));
var _time = _interopRequireDefault(require("./category/time"));
var _linear = _interopRequireDefault(require("./continuous/linear"));
var _log = _interopRequireDefault(require("./continuous/log"));
var _pow = _interopRequireDefault(require("./continuous/pow"));
var _time2 = _interopRequireDefault(require("./continuous/time"));
var _quantize = _interopRequireDefault(require("./continuous/quantize"));
var _quantile = _interopRequireDefault(require("./continuous/quantile"));
var _factory = require("./factory");
var _index = _interopRequireDefault(require("./identity/index"));
var _index2 = require("./tick-method/index");
